### PR TITLE
Changed category not defined error to warning in ElectronMVA

### PR DESCRIPTION
**Pull request by guitargeek**
Hello, unfortunately it's me again.

There were again some errors in the most recent IB due to the new electron MVA, related to the new category check I introduced [1]. I was unable to reproduce these errors locally so far*, but we can see that this problem affects not all electrons in the workflow. It crashes after a few 100 events.

By testing 1329.1 locally (which succeeded locally and failed in the IB), I observed that there were some electrons below 5 GeV in the crashing event, for which the MVA value gets computed. This is not foreseen by egamma, and if the category errors are exclusively on these very low pt electrons, everything is fine.

Hence I propose the following fix: let's change the _category not defined error_ to a warning that spits out pt and eta and just return -999 for the MVA if the category is not defined. We can check later in the logs if any electron in the kinematic region that the ID is designed for (pt > 5 GeV, |eta| < 2.5) failed, keeping an eye on the previously failing workflows which were:

1325.6
1325.8
1329.1
16.0 
134.903
136.777
136.797
10001.0

All due to the _category not defined_ error. The last one was actually a workflow that cms-bot successfully tested here at the end: https://github.com/cms-sw/cmssw/pull/23700. 

[1] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2.cc#L151

\* 1325.6 1325.8 and 16.0 gave me DAS errors, 1329.1 succeeded and for the others I didn't have time so far. Working on it.

### CODE REVIEW

File: RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2.cc [hunk 1]
```diff
@@ -104,6 +104,9 @@ float ElectronMVAEstimatorRun2::
 mvaValue( const edm::Ptr<reco::Candidate>& candPtr, const edm::EventBase & iEvent) const {
 
   const int iCategory = findCategory( candPtr );
+
+  if (iCategory < 0) return 0;
```
Comment (perrotta) [inline RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2.cc]: Normally, in case of bad  inputs a meaningless mva value is returned (e.g. -9999)
Here you return a 0: is it coherent with the logic of the algorithm?

**Comment by guitargeek** [comment]
It's not possible to return a meaningless MVA value at this stage, because it get logistically transformed afterwards [1], pushing large negative values to -1 and large positive values to 1.

To return 0 makes sense in the way that we don't know anything about the electron, it's right between background (-1) and signal (1). But I see your point: it's probably better to put a value such that these electrons are not being used. I'll put -999, which then gets transformed to -1 and then they are always considered as fakes. If one wants to check which electrons were affected by the undefined category, one can always look at the category value map, which stores the -1.

[1] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaTools/interface/MVAValueMapProducer.h#L146

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
abort tests

**Comment by perrotta** [comment]
please abort tests

**Comment by perrotta** [comment]
please test workflow 4,78,15.0,136.727,11626.0

**Comment by perrotta** [comment]
@guitargeek 

If having undefined MVA for electrons outside the egamma acceptance region is acceptable and understood (by the way: it is not clear to me why that undefined category is obtained for them: is the origin understood?), the LogWarning should only be issued for electrons with  (pt > 5 GeV, |eta| < 2.5) , in order not to fload the logs with useless messages from low pt electrons.

Please, let the tests finish before modifying anything, so that we can see if anything gets printed in the outputs. But feel free of answering my question above in the meanwhile :-)

**Comment by guitargeek** [comment]
Well the categories are defined in such a way that even the electrons outside the egamma acceptance get assigned to a category [1], and in my local tests of 1329.1 everything went fine. So the logs won't be written unless there is this problem we don't yet understand which causes all category formulas to return false.

That this caused by the pt < 5 GeV electrons was just my initial guess, since I saw some of them in the event 404 that crashed this workflow in the IB. My reasoning was that maybe some of the pt < 5 GeV electrons don't have an embedded supercluster, causing the formula to fail. But the truth is: I have no solid idea and it's hard to debug since I can't reproduce the crashes locally. It's just strange that everything else works for the crashing electrons, including the casting to the gsfElectron [2]! And then, all category formulas return 0 even though they are defined in such a way that always one should be true...

It's also not clear to me why everything worked before [3], were the categories were found using the same pt and superCluster.eta but just not with the cut parser. I'm investigating more in the afternoon.

[1] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_tools.py#L29
[2] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2.cc#L139
[3] https://github.com/cms-sw/cmssw/blob/CMSSW_9_4_X/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Fall17.cc#L213

**Comment by perrotta** [comment]
Ok, I see the point.

Therefore:
- a category must necessarily be assigned by construction
- if it is not assigned there is a bug somewhere
- this PR doesn't fix the bug,  but it helps you debugging the issue and find it

Please @guitargeek be fast with your debug, otherwise also #23700 seriously risk to be reverted, as we are too much close to the 10_2_0 deadline

**Comment by guitargeek** [comment]
Ok I'm doing my best! I now ran the workflows 134.903, 136.777 and 136.797 from the IB locally: no crashes again. It's hard to debug something that only crashes in the IB tests. Where are the root files from the IB workflows stored? Maybe I can get a hint by studying them.

**Comment by guitargeek** [comment]
Interesting, how can I do that? Or is this better done by experts?

**Comment by davidlange6** [comment]
The simplest approach is to run

valgrind cmsRun ....

(and wait..)

**Comment by guitargeek** [comment]
Ok thanks I did this with the cmsDriver.py command that I got from the workflow for step3. While it's running I'm investigating if the messed up categorization was already an issue before, maybe Fabio won't revert if I didn't make things worse at least...

**Comment by guitargeek** [comment]
Sorry to bother you but I have trouble using valgrind. If I just type `valgrind` in front of the cmsDriver command for step3, it outputs just a few lines in the beginning and then everything goes just normal without any valgrind output:

```
==31823== Memcheck, a memory error detector
==31823== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==31823== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==31823== Command: /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc630/cms/cmssw/CMSSW_10_2_X_2018-07-04-1900/bin/slc6_amd64_gcc630/cmsD
     river.py step3 --conditions auto:run2_data_promptlike -s RAW2DIGI,L1Reco,RECO,SKIM:TopMuEG,EI,PAT,ALCA:SiStripCalZeroBias+SiStripC
     alMinBias+TkAlMinBias+EcalESAlign,DQM:@standardDQM+@ExtraHLT+@miniAODDQM --runUnscheduled --process reRECO --data --era Run2_2017 
     --eventcontent RECO,MINIAOD,DQM --hltProcess reHLT --scenario pp --datatier RECO,MINIAOD,DQMIO --customise Configuration/DataProce
     ssing/RecoTLR.customisePostEra_Run2_2017 -n 100 --filein file:step2.root --fileout file:step3.root
==31823==
```

If I try to use the cmsPerfSuite.py which I read about in this TWiki [1], it just complains:

```
Error: An environment variable either SCRAM_ARCH, CMSSW_VERSION, HOST or USER is not available.
```

I'm trying to figure it out but I don't know if I can get to a conclusion in time for 10_2_0. I prepared a branch that reverts the new `findCategory()` with the cut parser [2]. Maybe I should just implement that in this PR and look into the actual problem another time when there is no release deadline. What do you think?

[1] https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuidePerformanceSuite#Introduction
[2] https://github.com/cms-sw/cmssw/compare/master...guitargeek:revert_cut_parser_categories

**Comment by guitargeek** [comment]
Ah yes, and when I do...

```
valgrind cmsRun step3_RAW2DIGI_L1Reco_RECO_SKIM_EI_PAT_ALCA_DQM.py
```
...I get a segmentation violation after 2 seconds and it quits.

**Comment by davidlange6** [comment]
i realize we have a conflict between valgrind and jemalloc in 102x - you can use cmsRunGlibC instead of cmsRun to avoid this conflict.

**Comment by fabiocos** [comment]
@guitargeek for 1325.6 I managed to reproduce the crash, use the --ibeos for runTheMatrix.py

**Comment by fabiocos** [comment]
@guitargeek I will not revert the original PR (and the corresponding external) for the time being, to simplify your debugging. But I agree with Andrea that this PR is not the solution, at most a debugging tool. Unfortunately, if we do not converge on a real solution by the week-end, I would say that we will have to reschedule this development for 10_3_X

**Comment by guitargeek** [comment]
Thanks Fabio for the clarification and for not reverting for now. I will do my best to find a solution by tomorrow. The valgrind works now with Davids hint, and I hope the error will also show up in my 1325.6 workflow. Once I can reproduce the error, things should be simpler!

**Comment by guitargeek** [comment]
@fabiocos, can you please tell me exactly how you could reproduce the errors? I'm currently trying it with `--ibeos` on lxplus with CMSSW_10_2_X_2018-07-05-1100, but all the workflows I mention in the initial post succeeded.

Here is some first valgrind output: https://rembserj.web.cern.ch/rembserj/dumps/valgrind_1.log.

I ran step3 of workflow 136.797 with `valgrind --leak-check=yes cmsRunGlibC`. There is a pretty nasty leak in the GBRForestTools [1] due to a returned buffer created with `malloc` here [2]. You think this could be an issue? However it's easy to fix with a corresponding `free` in the GBRForestTools. Apart from this issue, keywords corresponding to my new classes like `ElectronMVAEstimatorRun2`, `MVAVariableManager` and `MVAValueMapProducer` don't show up in the report.

There are a few occurrences of the word "Electron" though, especially related to the `Conditional jump or move depends on uninitialised value(s)` errors. I'm running valgrind again with `--track-origins=yes` to understand these better.

tl;dr - I still have no idea what's going on...

[1] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaTools/src/GBRForestTools.cc#L65
[2] https://github.com/cms-sw/cmssw/blob/master/CommonTools/Utils/src/TMVAZipReader.cc#L39

**Comment by perrotta** [comment]
please test workflow 1325.7,1325.9,136.7722,4.27,136.727,10002.0

**Comment by guitargeek** [comment]
I looked at the logs from the workflows that Andrea explicitly asked for: not a single occurrence of a `category not defined for particle` warning...

**Comment by perrotta** [comment]
please test workflow 1325.7,1325.9,136.7722,4.27,136.727,10002.0

**Comment by perrotta** [comment]
It is weird that it cannot be reproduced by jenkins tests of this PR, while it keeps being hit in the IB tests.

Me too, I made several attempts with the workflows which are failing in the IBs, and I was never able to see a single crash. Apparently they do only crash in IB RelVals... 

Looking at the code,and according to me, the only possibility that [1] never returns "true" is when either pt or eta are nan's:. (Therefore, an exception there is not recommended, and it should be removed anyhow, as it is done here). It could be useful to see in the IB logs if it is the case (and afterwards to understand why that does not happen in our private tests with the same workflows, nor in these jenkins tests).

As a last attempt, I would try to merge this PR and see in the IB logs what is output by the LogWarning. But I understand that we are quite late for it, and I let @fabiocos decide whether it is still worth doing so.

[1] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_tools.py#L29

**Comment by perrotta** [comment]
+1
- The issue with #23473 only shows up in the IB RelVals: not in the jenkins nor in the private tests
- At this point, either #23473 is reverted, or this one is also merged in the iB to help understanding the origin of the problem: this signature for reco is intended to allow that possible merge, in case we decide for that

**Comment by smuzaffar** [comment]
@perrotta , note that in IB we run multi-threaded jobs while PR tests are single threaded.

**Comment by fabiocos** [comment]
+1 

in my standalone tests I usually run with -t 4. Let's try this PR

**Comment by guitargeek** [comment]
Thanks @smuzaffar and @fabiocos, I'm trying some workflows with `-t 4` now.

Thinking a big ahead, what could be the problem with the ElectronID in multi threaded mode? In my opinion it's related to the new implementation of `findCategory` [1] with the cut parser, because all other changes in the electron MVA were also implemented in the photon MVA which never crashes. This function is used in the `ElectronMVAEstimatorRun2` and `MVAValueMapProducer` classes.

Maybe I could start by making the `MVAValueMapProducer` a global module instead of a steam module? It doesn't need to be, and I would guess stream modules are more error prone?

[1] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2.cc#L140
[2] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaTools/interface/MVAValueMapProducer.h

**Comment by guitargeek** [comment]
So we observe segfaults now... In this case I can understand that a revert of https://github.com/cms-sw/cmssw/pull/23746 and https://github.com/cms-sw/cmssw/pull/23700 would be a good idea, because I still can't reproduce the crashes, even with `t -4`. Maybe they don't appear every time.

In any case I wanted to give a little idea for the revert, @fabiocos . If you change every appearance of `.xml` to `.xml.gz` in all these files here, you don't have to revert the external:

* https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Fall17_iso_V1_cff.py
* https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Fall17_noIso_V1_cff.py
* https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Spring15_25ns_Trig_V1_cff.py
* https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Spring15_25ns_nonTrig_V1_cff.py
* https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Spring15_50ns_Trig_V1_cff.py
* https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Spring16_GeneralPurpose_V1_cff.py
* https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Spring16_HZZ_V1_cff.py

In other words all the *mvaElectronID_*.py here [1]. It would be also good to keep the GBRForestTools as they are now because of the memory-leak fix:

* https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaTools/src/GBRForestTools.cc
* https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaTools/interface/GBRForestTools.h

[1] https://github.com/cms-sw/cmssw/tree/master/RecoEgamma/ElectronIdentification/python/Identification

**Comment by perrotta** [comment]
In the meanwhile, the IB with this PR was produced. Now the logWarning message says (e.g. wf 136.7721):

```
Begin processing the 12th record. Run 283877, Event 73383640, LumiSection 43 on stream 2 at 09-Jul-2018 05:18:54.295 CEST
%MSG-w MVA warning:  ElectronMVAValueMapProducer:electronMVAValueMapProducer  09-Jul-2018 05:18:55 CEST Run: 283877 Event: 73383640
category not defined for particle with pt 6.53122 GeV, eta -0.533662 in ElectronMVAEstimatorRun2Spring16GeneralPurposeV1
```

which is weird, because that pt/eta pair must have a category assigned.

And right after, the job crashes as well with a segmentation violation, see https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc630/CMSSW_10_2_X_2018-07-08-2300/pyRelValMatrixLogs/run/136.7721_RunJetHT2016H_reminiaod+RunJetHT2016H_reminiaod+REMINIAOD_data2016+HARVESTDR2_REMINIAOD_data2016/step2_RunJetHT2016H_reminiaod+RunJetHT2016H_reminiaod+REMINIAOD_data2016+HARVESTDR2_REMINIAOD_data2016.log

I'm sorry, but I think that this issue must be sorted out before the electron MVA ID can be accepted in the release. Unless @guitargeek can ultra-quickly get some clever idea out of that log output, I see no other option than reverting the two PRs and the external

**Comment by guitargeek** [comment]
No, I don't understand it either. It seems to be some deeper problem, probably related to multithreading, that I can't sort out ultra-quick.

We should revert, I agree... See my post above about how reverting the external can be avoided.

I guess for the ID I will set up a recipe in the egamma Twiki and with the refactored ID implementation I'll come back here another time.



On July 9, 2018 8:31:00 AM GMT+02:00, perrotta <notifications@github.com> wrote:
>In the meanwhile, the IB with this PR was produced. Now the logWarning
>message says (e.g. wf 136.7721):
>
>```
>Begin processing the 12th record. Run 283877, Event 73383640,
>LumiSection 43 on stream 2 at 09-Jul-2018 05:18:54.295 CEST
>%MSG-w MVA warning: 
>ElectronMVAValueMapProducer:electronMVAValueMapProducer  09-Jul-2018
>05:18:55 CEST Run: 283877 Event: 73383640
>category not defined for particle with pt 6.53122 GeV, eta -0.533662 in
>ElectronMVAEstimatorRun2Spring16GeneralPurposeV1
>```
>
>which is weird, because that pt/eta pair must have a category assigned.
>
>And right after, the job crashes as well with a segmentation violation,
>see
>https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc630/CMSSW_10_2_X_2018-07-08-2300/pyRelValMatrixLogs/run/136.7721_RunJetHT2016H_reminiaod+RunJetHT2016H_reminiaod+REMINIAOD_data2016+HARVESTDR2_REMINIAOD_data2016/step2_RunJetHT2016H_reminiaod+RunJetHT2016H_reminiaod+REMINIAOD_data2016+HARVESTDR2_REMINIAOD_data2016.log
>
>I'm sorry, but I think that this issue must be sorted out before the
>electron MVA ID can be accepted in the release. Unless @guitargeek can
>ultra-quickly get some clever idea out of that log output, I see no
>other option than reverting the two PRs and the external
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/cms-sw/cmssw/pull/23746#issuecomment-403373527

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
